prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -22,54 +26,131 @@ def floatify(value, default=float("nan")):
+        if not isinstance(generator["xyz"], list):
             return pd.NA
-    distance_to_subs = matching_subs.apply(
-        lambda x: haversine((x.lat, x.lon), (generator.lat, generator.lon)),
+        if pd.isnull(generator["voltage_class"]) or generator["Pmax"] < 100:
+            grouper_key = generator["interconnect"]
+        else:
+            grouper_key = (generator["interconnect"], generator["voltage_class"])
+        chord_dist, array_index = voltage_trees[grouper_key].query(generator["xyz"])
+        sub_id = subs_voltage_lookup[grouper_key][array_index]
+        # Translate chord distance (unit circle) to great circle distance (miles)
+        dist_in_miles = 3963 * 2 * asin(chord_dist / 2)  # use 3963 mi as earth radius
+        return pd.Series({"dist": dist_in_miles, "sub_id": sub_id})
+
+    def classify_voltages(voltage, voltage_ranges):
+        for v_range, bounds in voltage_ranges.items():
+            if bounds["min"] <= voltage <= bounds["max"]:
+                return v_range
+        return float("nan")
+
+    voltage_ranges = {
+        "under 100": {"min": 0, "max": 99},
+        "100-161": {"min": 100, "max": 161},
+        "220-287": {"min": 220, "max": 287},
+        "345": {"min": 345, "max": 345},
+        "500": {"min": 500, "max": 500},
+        "735 and above": {"min": 735, "max": float("inf")},
+    }
+
+    # Translate lat/lon to 3D positions (assume spherical earth, origin at center)
+    substations_with_xyz = substations.assign(
+        xyz=substations.apply(lambda x: latlon_to_xyz(x["lat"], x["lon"]), axis=1)
+    )
+    generators_with_xyz = generators.assign(
+        xyz=generators.apply(
+            lambda x: (
+                pd.NA
+                if pd.isna(x["lat"]) or pd.isna(x["lon"])
+                else latlon_to_xyz(x["lat"], x["lon"])
+            ),
+            axis=1,
+        )
+    )
+
+    # Bin voltages into broad classes
+    substations_with_xyz["voltage_class"] = substations["MAX_VOLT"].map(
+        lambda x: classify_voltages(x, voltage_ranges)
+    )
+    generators_with_xyz["voltage_class"] = generators["Grid Voltage (kV)"].map(
+        lambda x: classify_voltages(x, voltage_ranges)
+    )
+
+    # Group substations by voltage to build KDTrees
+    subs_voltage_lookup = {
+        (interconnect, voltage_level): substations_with_xyz.loc[
+            (substations_with_xyz["interconnect"] == interconnect)
+            & (substations_with_xyz["MAX_VOLT"] >= voltage_range["min"])
+        ].index
+        for interconnect in generators["interconnect"].unique()
+        for voltage_level, voltage_range in voltage_ranges.items()
+    }
+    # Group substations by ZIP code for a fallback for generators without coordinates
+    subs_zip_groupby = substations_with_xyz.groupby(["interconnect", "ZIP"])
+
+    # Create a KDTree for each combination of state and interconnect
+    voltage_trees = {
+        key: KDTree(np.array(substations_with_xyz.loc[sub_ids, "xyz"].tolist()))
```

Comment:
Are we using `numpy.ndarray` for performance concern? If so, maybe we can do `substations_with_xyz.loc[sub_ids, "xyz"].to_numpy()` directly?

Reply:
If I remember correctly, each entry in the `"xyz"` column is a list/tuple, so calling `.to_numpy()` directly gives us a 1D array of list/tuple objects, not the 2D array that we really want. I'll double-check to be sure. I agree that it's ugly.

Reply:
Yeah, I tried it and found the issue: if value of each cell is list/tuple, `.to_numpy()` will give us a 1D array with entries to be a weird format, i.e. `array([list([a, b, c]), list([d, e, f])])`. Here is one solution I found from google which works for values being same length across the column: `np.vstack(substations_with_xyz.loc[sub_ids, "xyz"])`.

Reply:
Nice! Done.